### ISSUE TITLE
First Class Authorization and Token Management Support

### ISSUE THREAD

**Issue by pdevito3** [dotnet/aspnetcore#38268]
## Preface

Given the call for .NET 7 features in #37373, I wanted to toss this out there. I don't want to beat a dead horse or start another controversial ticket, but this a common request as one of the biggest gaps and barriers to entry in the .NET ecosystem. The closest recent ticket I've found for this is #26594 which was locked (I'd think for good reason). 

I'm opening this ticket to give us a location to vote on and track this over time in, what I hope can be, a more civil discussion than prior forums.

## The Issue

As #26594 mentions, other frameworks (e.g. Java and PHP/Laravel ecosystems) have pretty plug and play solutions for token management services. The .NET community doesn't have this and relies on external options to put together a fully functional solution.

This third party dependency adds a lot of overhead, risk, and cost to current community members and is a major detractor for newcomers comparing .NET to other options.

### What Exists Now

MS Identity lays a foundation and is critical to our success here, but still has gaps (this is why Identity Server was created in the first place). AAD fills these gaps, but is paid, brings in vendor lock, and is outside of our .NET realm. Solid third party options like [Duende](https://duendesoftware.com/) and [OpenIddict](https://github.com/openiddict/openiddict-core) exist, but bring a dependency outside of our Core.

Duende is what I generally see as the recommended option for us right now and for good reason. I am a huge fan and will be using it in my projects for the foreseeable future, even with the eventual cost aspect. I personally applaud them for doing what they needed to to support their project.

Regardless of how your feel about Duende though, it still requires us to rely on an external service to fill a critical need for apps built using .NET. 

### Summary

The fact remains that there is no plug and play solution for the .NET community at this point. This adds a ton of overhead and cost for existing community members and is a major detractor/barrier to entry for devs looking to join the ecosystem.

## Potential Barriers

I believe the major hurdles to this that I've seen mentioned are:

1. The desire to, as the community has requested, not interfere with the OSS community

   > I believe OpenIddict is the last man standing on this front and, while I like the sentiment here, even if the need is filled by OSS, it feels wrong to me that we as a community need to go outside our ecosystem to fill such a critical, common, and complex requirement.  

2. The MS experts for this domain live in the AAD team, not the .NET team and that AAD is the recommended 1st party support at this point.

   > I get this, but at the same time, most of us aren't security experts either. If .NET doesn't give us guidance other than point us to paid services, that puts us as a deficit when building apps and honestly puts .NET a major disadvantage when compared to other frameworks. I would hope that the team could bring on some experts to support this need and benefit the entire platform.

3. This is a large effort to build and maintain and carries a lot of risk

   > I hear you guys on this one. This would be a huge lift that would take a lot of resources and commitment over time, but the fact remains that this is a gap in our community and it would be hugely beneficial to support something like this. 

I'm sure I'm missing plenty of other hurdles as well. There's plenty to overcome and this would be a huge lift, but the gains would be substantial as well. This may even be too much to overcome for the time being, but I'm hoping this ticket can act as a location to establish the impact it might have on the community and be used to help prioritize/deprioritize this effort.

## TLDR

While there are solid 3rd party solutions out there, the lack of easy and reliable token service support in .NET is a major gap in our ecosystem. This would be a big lift, but if we can fill this gap, it would be a huge value add to the existing community and would break down one of the few remaining major barriers to entry to .NET.

**Comment by pdevito3** [comment]
@blowdart i'm assuming you'd own this? 

again, don't mean to stir the pot, just want to put it out there for voting and the community to hopefully have a reasonable conversation about... though i know that last part has previously been a challenge, so I completely get if it needs to be shut down.

**Comment by blowdart** [comment]
Our position hasn't changed - we don't have plans for this. 

Obviously we'd love you to use azure ad but _any_ cloud hosted identity solution is going to be better than anything we'd produce and much easier to secure. If you have reasons not to use the cloud then either identity server or open iddict are specialised, focussed, solutions which again are better than anything we'd do.

**Comment by pdevito3** [comment]
> We can't just bring in experts, writing such a server would be an ongoing commitment, it's not a one and done thing.

Yup, didn't mean to imply they wouldn't be. Would definitely need to be a new, ongoing part of the team. 

I hear you on cloud options being the best and the existing 3rd party options being the next best thing. I know there's huge amounts of complexity to account for when setting up something like this. There's whole companies built around this and it is very much a non trivial ask.

With that said, when other frameworks find a way to have built in solutions and we don't, it's a pretty big detractor for us. Especially when we're trying to expand the community. 

I understand why you guys stand where you do and completely respect it, but wanted to at least keep the conversation alive.

**Comment by davidfowl** [comment]
> Solid third party options like Duende and OpenIddict exist, but bring a dependency outside of our Core.

Would more docs with both of these solutions help?

**Comment by pdevito3** [comment]
While we'd still be dependent on third party solutions, I personally think that would be a big step forward.

We should make it easy and approachable for devs to set up auth in their projects without having to reach for something like Auth0 or some other expensive OOTB service. Some thoughts on topics to cover: 

- The latest auth standards. i.e the security community is approaching OAuth 2.1 and the only two recommended flows are code w/ PKCE or client credentials depending on whether you’re interactive or not respectively. 
- This is how you can set up a duende server. [Here](https://www.scottbrady91.com/identity-server/getting-started-with-identityserver-4) is an example of at least some of what could be covered
- This is how you can add permissions, rules, and policies
- This is how you can integrate in API and Swagger clients, for example
- This is how you can integrate identity
- This is how you can set up a BFF if you’re working with a spa
- Here are useful methods for common boilerplate features like password reset, password guessing, pwn prevention, account verification, MFA, etc. 

Some of this exists already and there are probably several other items I’m missing as well, but regardless I think the end goal is that *devs should be able to walk away from the docs and be able to set up a functional auth server of their own.* Maybe that includes a supplemental example reference repo/template repo? 

Obviously you can’t cover anywhere near every nuance, but having a single high quality resource that shows how the creators of the .NET framework recommend that we set up quality auth server base inside their own ecosystem would be huge. Right now we are left to our own devices.

**Comment by davidfowl** [comment]
I agree. I would lean heavily towards spending energy here documenting and possibly even auth streamlining common scenarios with openiddict and duende.

**Comment by blowdart** [comment]
Of course the risk here is our docs get out of date as Identity Server or Open Iddict release new versions.

**Comment by davidfowl** [comment]
I'd be willing to bet that's a low risk in the grand scheme of things.

**Comment by robloo** [comment]
This msftbot is seems to have close this way too early. The discussion does not seem to have completed and the issue isn't resolved. The Microsoft team is throwing out a lot of good ideas. Then locking issues like this is almost a sin.

**Comment by pdevito3** [comment]
Wanted to point to [this thread](https://twitter.com/pdevito3/status/1540874677991706624?s=21&t=r4DD2rfRwKCHhFN7ewLnnQ) as I thought it was an interesting discussion around this topic.

Essentially, the thought is that having .NET ship an OIDC or OAuth2 server implementation that competes with others in the ecosystem would kill them which I don't think anyone wants. 

Instead, I'd propose that if we really are leaning on Duende and OpenIDDict as the offerings we need to be using in our ecosystem, I think it would benefit the community if MS  leans into this more and provides a lot more guidance on how we can do this. More docs, templates, resources, courses, etc. 

I know you already mentioned this above @davidfowl, so idk if that continues here or elsewhere (or isn't even on the table 🤷), but making this as approachable and as easy as we can will really elevate the .NET community.
